sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/ForeachWriterSuite.scala

```diff
@@ -286,6 +293,71 @@ object ForeachWriterSuite {
   }
 }
 
+class ForeachWriterAbortSuite extends StreamTest with SharedSQLContext with BeforeAndAfter {
```

Comment:
I think this complicated test suite is not necessary. Could you take a look at the `foreach with error` test in `ForeachWriterSuite` and see if you can write a similar one, such as calling `df.map(_ / 0)` to trigger an error?

Reply:
I guess this triggers different situation, while `df.map(_ / 0)` just fails the processing, this would fail "between" writing has been successful and committing is happening.

IMHO we even need to have a test which `commit` is failing and `abort` is being called, but python test is covering (the reason of hang) so we might be OK to  rely on it.

Reply:
in the case of `foreach with error` test, the abort is being called but it will not close the writer (since it is already closed) and the abort exception will not be thrown
i wanted to create a test that will simulate a use case in which the abort is called when the writer has not been closed yet

Reply:
> IMHO we even need to have a test which commit is failing and abort is being called
+1

Reply:
@zsxwing what do you think? can we remain with the existing tests?

Reply:
> i wanted to create a test that will simulate a use case in which the abort is called when the writer has not been closed yet

`df.map(_ / 0)` will trigger the above behavior. Right?

Reply:
Here is the regression test I meant:
```Scala
  testQuietly("foreach with error not caused by ForeachWriter") {
    withTempDir { checkpointDir =>
      val input = MemoryStream[Int]
      val query = input.toDS().repartition(1).map(_ / 0).writeStream
        .option("checkpointLocation", checkpointDir.getCanonicalPath)
        .foreach(new TestForeachWriter)
        .start()
      input.addData(1, 2, 3, 4)

      val e = intercept[StreamingQueryException] {
        query.processAllAvailable()
      }
      assert(e.getCause.isInstanceOf[SparkException])
      assert(e.getCause.getCause.getCause.getMessage === "/ by zero")
      assert(query.isActive === false)

      val allEvents = ForeachWriterSuite.allEvents()
      assert(allEvents.size === 1)
      assert(allEvents(0)(0) === ForeachWriterSuite.Open(partition = 0, version = 0))
      // `close` should be called with the error
      val errorEvent = allEvents(0)(1).asInstanceOf[ForeachWriterSuite.Close]
      assert(errorEvent.error.get.isInstanceOf[SparkException])
      assert(errorEvent.error.get.getMessage ===
        "Foreach writer has been aborted due to a task failure")
    }
  }
```

Reply:
that will do! thanks